src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -139,17 +140,33 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @Ch
             BitbucketAuthenticator authenticator = AuthenticationTokens.convert(BitbucketAuthenticator.authenticationContext(serverUrl), credentials);
 
             BitbucketApi apiClient = BitbucketApiFactory.newInstance(serverUrl, authenticator, owner, repository);
-            String ref;
+            String ref = null;
+
             if (head instanceof BranchSCMHead) {
                 ref = head.getName();
-            } else if (head instanceof PullRequestSCMHead) {
+            }
+            if (head instanceof PullRequestSCMHead) {
+                // working on a pull request - can be either "HEAD" or "MERGE"
                 PullRequestSCMHead pr = (PullRequestSCMHead) head;
-                if (!(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) && pr.getRepository() != null) {
-                    return new BitbucketSCMFileSystem(apiClient, pr.getOriginName(), rev);
+                if (pr.getRepository() == null) { // not clear when this happens
+                    return null;
+                }
+
+                if (apiClient instanceof BitbucketCloudApiClient) {
+                    if (pr.getCheckoutStrategy() != ChangeRequestCheckoutStrategy.MERGE) {
+                        return new BitbucketSCMFileSystem(apiClient, pr.getOriginName(), rev);
+                    }
+                    return null; // TODO support merge revisions somehow for cloud
```

Comment:
Perhaps link to https://bitbucket.org/site/master/issues/5814/refify-pull-requests-by-making-them-a-ref